Marlin/src/module/temperature.h

```diff
@@ -794,14 +798,8 @@ class Temperature {
         #endif
       );
       #if ENABLED(AUTO_REPORT_TEMPERATURES)
-        static uint8_t auto_report_temp_interval;
-        static millis_t next_temp_report_ms;
-        static void auto_report_temperatures();
-        static inline void set_auto_report_interval(uint8_t v) {
-          NOMORE(v, 60);
-          auto_report_temp_interval = v;
-          next_temp_report_ms = millis() + 1000UL * v;
-        }
+        class AutoReportTemp : public AutoReporter<SERIAL_ALL> { void auto_report(); };
```

Comment:
`auto_report` is not virtual so this is never called when manipulating the base class.

Reply:
If you want to avoid the cost of a virtual table (since the instance is known at compile time), you can use CRTP here:
```
template <class Child>
struct AutoReporter 
{
[...]
inline void auto_report() { static_cast<Child*>(this)->report(); } // Call the child's version. Does not need to be called auto_report
[...]
};

// You need a default implementation that does not report.
struct NoAutoReport : public AutoReporter<NoAutoReport> { void report() {} };

// Then subclass like this:
struct AutoReportTemp : public AutoReporter<AutoReportTemp> { void report() { ... } };

``` 
But this means you'll have to typedef the actual class to use since the type will change on each platform:
```
// In some common header, select the type to use based on compile time information:
#ifdef HAS_MULTI_SERIAL
  typedef AutoReportMulti AutoReport;
#else if ...
  typedef NoAutoReport AutoReport;
[...]
#endif
// Ok, let's declare the reporter instance
extern AutoReport auto_reporter;
```